### ISSUE TITLE
parse errors in *x

### ISSUE THREAD

**Issue by Smattr** [rems-project/sail#45]
I don't know if sail2 is meant to be considered unstable at this point, but using commit 57443173923e87f33713c99dbab9eba7e3db0660 I'm experiencing the following:

```sh
$ cat test.sail
default Order dec
$include <prelude.sail>

infix 9 :=

val main : unit -> unit effect {wreg}
function main() = {
  (*x) := 0xcafebeef
}
$ sail test.sail
Internal error: Unreachable code (at "initial_check.ml" line 414):
[test.sail]:8:3-5
8 |  (*x) := 0xcafebeef
  |   ^^
  | Unparseable construct in to_ast_exp
  |
  | Please report this as an issue on GitHub at https://github.com/rems-project/sail/issues
```

The spec is kind of nonsense, but I was not expecting an ICE.

Apart from this, my experience with Sail has so far been very positive. Very nice work :)

**Comment by Alasdair** [comment]
Currently (*x) only works on the left hand side of an expression as an assignment to a register reference. The line of code in question should have raised a syntax error rather than an internal error in this case.

However making it desugar to a register dereference is probably a better thing to do, rather than a syntax error, so the following would parse and typecheck:

```
default Order dec
$include <prelude.sail>

infix 9 :=

val operator := : (bits(32), bits(32)) -> unit

register R : bits(32)

val main : unit -> unit effect {rreg}

function main() = {
  let x = ref R;
  (*x) := 0xcafebeef
}
```
(defining your own operator called := is not really recommended because this means something quite different to how it looks!)

I think the syntax of Sail can probably be considered somewhat stable at this point, but I don't think we have enough users that we'd 100% want to restrict ourselves from making breaking changes if we felt it'd improve the language if that makes sense.

**Comment by Smattr** [comment]
Ah, yes, `*x` only being valid on register references was what I concluded too by looking at other examples. I initially only wrote the spec with `:=` as a typo, but it didn't occur to me to try to further minimise the reproducer by removing that.

**Comment by Alasdair** [comment]
Should be fixed in 3677cfc13e

**Comment by Smattr** [comment]
Thanks! I'm surprised `__deref` had to become a magical thing the ML innards recognises.
